File: ProgrammingIdeas/Activities/IdeaListActivity.cs
using Android.App;
using Android.Content;
using Android.OS;
using Android.Support.V7.Widget;
using Android.Views;
using Android.Widget;
using ProgrammingIdeas.Adapters;
using ProgrammingIdeas.Helpers;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using PopupMenu = Android.Support.V7.Widget.PopupMenu;
using Android.Support.Design.Widget;

namespace ProgrammingIdeas.Activities
{
	[Activity(Label = "ItemActivity", Theme = "@style/AppTheme")]
	public class IdeaListActivity : BaseActivity, PopupMenu.IOnMenuItemClickListener
	{
		private RecyclerView recyclerView;
		private RecyclerView.LayoutManager manager;
		private IdeaListAdapter adapter;
		private List<Category> allCategories = new List<Category>();
		private List<Idea> ideasList;
		private ProgressBar progressBar;
		private string ideasdb = Path.Combine(Global.APP_PATH, "ideasdb");

		public override int LayoutResource => Resource.Layout.idealistactivity;

		public override bool HomeAsUpEnabled => true;

		protected override void OnCreate(Bundle savedInstanceState)
		{
			base.OnCreate(savedInstanceState);
			manager = new LinearLayoutManager(this);
			recyclerView = FindViewById<RecyclerView>(Resource.Id.itemRecyclerView);
			progressBar = FindViewById<ProgressBar>(Resource.Id.completedIdeasBar);
			allCategories = Global.Categories;
			progressBar.Max = allCategories[Global.CategoryScrollPosition].Items.Count;
			SetupUI();
		}

		protected override void OnResume()
		{
			base.OnResume();
			if (Global.RefreshBookmarks)
				adapter.RefreshBookmarks();
		}

		public override bool OnCreateOptionsMenu(IMenu menu)
		{
			MenuInflater.Inflate(Resource.Menu.idea_list_menu, menu);
			return base.OnCreateOptionsMenu(menu);
		}

		public override bool OnOptionsItemSelected(IMenuItem item)
		{
			switch (item.ItemId)
			{
				case Resource.Id.sortIdeas:
					var sortAnchor = FindViewById(Resource.Id.sortIdeas);
					ShowSortPopup(sortAnchor);
					return true;
			}
			return base.OnOptionsItemSelected(item);
		}

		private void ShowSortPopup(View sortAnchor)
		{
			var popup = new PopupMenu(this, sortAnchor);
			popup.MenuInflater.Inflate(Resource.Menu.idea_sort_menu, popup.Menu);
			popup.SetOnMenuItemClickListener(this);
			popup.Show();
		}

		private void ShowProgress()
		{
			var completedCount = allCategories[Global.CategoryScrollPosition].Items.FindAll(x => x.State == Status.Done).Count;
			progressBar.Progress = 0;
			progressBar.IncrementProgressBy(completedCount);
		}

		private void SetupUI()
		{
			var title = Global.Categories[Global.CategoryScrollPosition].CategoryLbl;
			ideasList = Global.Categories[Global.CategoryScrollPosition].Items;

			Title = title;
			recyclerView.SetLayoutManager(manager);
			adapter = new IdeaListAdapter(ideasList);
			adapter.ItemClick -= OnItemClick;
			adapter.ItemClick += OnItemClick;
			recyclerView.SetAdapter(adapter);
			manager.ScrollToPosition(Global.ItemScrollPosition);
			adapter.StateClicked -= Adapter_StateClicked;
			adapter.StateClicked += Adapter_StateClicked;

			ShowProgress();
		}

		void Adapter_StateClicked(string title, string state, int adapterPos)
		{
			if (ideasList != null && ideasList.Count != 0)
			{
				var changedItem = ideasList.ElementAt(adapterPos);
				if (changedItem != null)
				{
					changedItem.State = state;
					adapter.NotifyItemChanged(adapterPos);
					ShowProgress();
				}
			}
		}

		protected override void OnPause()
		{
			DBAssist.SerializeDBAsync(ideasdb, allCategories);
			base.OnPause();
		}

		private void OnItemClick(int position)
		{
			Global.ItemScrollPosition = position;
			StartActivity(new Intent(this, typeof(IdeaDetailsActivity)));
			OverridePendingTransition(Resource.Animation.push_left_in, Resource.Animation.push_left_out);
		}

		public bool OnMenuItemClick(IMenuItem item)
		{
			switch (item.ItemId)
			{
				case Resource.Id.sortByName:
					var sortedList = ideasList.OrderBy(x => x.Title).ToList();
					ideasList.Clear();
					ideasList.AddRange(sortedList);
					adapter.NotifyDataSetChanged();
					return true;

				case Resource.Id.sortByDifficulty:
					var sortedDiffList = ideasList.OrderBy(x => x.GetDifficultyId()).ToList();
					ideasList.Clear();
					ideasList.AddRange(sortedDiffList);
					adapter.NotifyDataSetChanged();
					return true;

				case Resource.Id.sortByStatus:
					var sortedStatusList = ideasList.OrderBy(x => x.GetStatusId()).ToList();
					ideasList.Clear();
					ideasList.AddRange(sortedStatusList);
					adapter.NotifyDataSetChanged();
					return true;

				default:
					return true;
			}
		}
	}
}

File: android/kotlin/app/src/main/java/com/alansa/ideabag2/views/IdeaListActivity.kt
package com.alansa.ideabag2.views

import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.PopupMenu
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.alansa.ideabag2.BaseActivity
import com.alansa.ideabag2.Global
import com.alansa.ideabag2.R
import com.alansa.ideabag2.adapters.IdeaListAdapter
import com.alansa.ideabag2.databinding.ActivityIdeaListBinding
import com.alansa.ideabag2.dialogs.SetProgressDialog
import com.alansa.ideabag2.extensions.empty
import com.alansa.ideabag2.utils.SortOption
import com.alansa.ideabag2.viewmodels.IdeaListViewModel
import kotlinx.android.synthetic.main.activity_idea_list.*

class IdeaListActivity : BaseActivity() {
    private lateinit var binding: ActivityIdeaListBinding
    private lateinit var viewmodel: IdeaListViewModel
    private lateinit var adapter: IdeaListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_idea_list)
        viewmodel = ViewModelProviders.of(this).get(IdeaListViewModel::class.java)
        binding.viewmodel = viewmodel

        setupToolbar()
        supportActionBar?.title = viewmodel.ideas[0].category

        setupList()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.idea_list_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.sortIdeas -> showSortIdeasPopup()
        }
        return true
    }

    private fun showSortIdeasPopup() {
        val anchor = findViewById<View>(R.id.sortIdeas)
        val popup = PopupMenu(this, anchor)
        popup.menuInflater.inflate(R.menu.idea_sort_menu, popup.menu)
        popup.setOnMenuItemClickListener {
            when (it?.itemId) {
                R.id.sortByDifficulty -> {
                    viewmodel.sortIdeas(SortOption.DIFFICULTY)
                    adapter.notifyDataSetChanged()
                }
                R.id.sortByName -> {
                    viewmodel.sortIdeas(SortOption.NAME)
                    adapter.notifyDataSetChanged()
                }
            }

            return@setOnMenuItemClickListener true
        }
        popup.show()
    }

    private fun setupList() {
        adapter = IdeaListAdapter(viewmodel.categoryId, viewmodel.ideas, { itemClicked(it) }, { itemLongClicked(it) })
        itemRecyclerView.layoutManager = LinearLayoutManager(this)
        itemRecyclerView.adapter = adapter
    }

    private fun itemLongClicked(position: Int) {
        var dialog = SetProgressDialog() { status ->
            viewmodel.setIdeaProgress(status, position + 1)
            adapter.notifyIdeaStatusChanged(position)
            Snackbar.make(itemRecyclerView, "Progress updated.", Snackbar.LENGTH_SHORT).show()
        }
        dialog.show(supportFragmentManager, String.empty)
    }

    private fun itemClicked(position: Int) {
        startActivity(Intent(this, IdeaDetailActivity::class.java))
        overridePendingTransition(R.anim.push_left_in, R.anim.push_left_out)
        Global.ideaClickIndex = position
    }

    override fun onResume() {
        super.onResume()
        adapter.refresh()
    }
}
